core/src/test/java/org/apache/iceberg/util/TestLocationUtil.java

```diff
@@ -61,4 +61,124 @@ public void testStripTrailingSlashWithInvalidPath() {
           () -> LocationUtil.stripTrailingSlash(invalidPath));
     }
   }
+
+  @Test
+  public void testPosixNormalizePathsWithSchemeAndAuthorityS3Style() {
+    Assert.assertEquals(
+        "Must work with a root authority",
+        "s3://bucket",
+        LocationUtil.posixNormalize("s3://bucket"));
+    Assert.assertEquals(
+        "Must remove / from the end of authority",
+        "s3://bucket",
+        LocationUtil.posixNormalize("s3://bucket/"));
+    Assert.assertEquals(
+        "Must remove / from the end of directory",
+        "s3://bucket/dir",
+        LocationUtil.posixNormalize("s3://bucket/dir/"));
+    Assert.assertEquals(
+        "Must change // to / to enforce posix path",
+        "s3://bucket/foo/bar",
+        LocationUtil.posixNormalize("s3://bucket/foo//bar"));
+    Assert.assertEquals(
+        "Must resolve .. to previous directory to enforce posix path",
+        "s3://bucket/bar",
+        LocationUtil.posixNormalize("s3://bucket/foo/../bar"));
+    Assert.assertEquals(
+        "Must resolve . to current directory to enforce posix path",
+        "s3://bucket/foo/bar",
+        LocationUtil.posixNormalize("s3://bucket/foo/.//bar"));
+  }
+
+  @Test
+  public void testPosixNormalizePathsWithSchemeAndAuthorityHdfsStyle() {
+    Assert.assertEquals(
+        "Must work with a root authority",
+        "hdfs://1.2.3.4",
+        LocationUtil.posixNormalize("hdfs://1.2.3.4"));
+    Assert.assertEquals(
+        "Must remove / from the end of authority",
+        "hdfs://1.2.3.4",
+        LocationUtil.posixNormalize("hdfs://1.2.3.4/"));
+    Assert.assertEquals(
+        "Must remove / from the end of directory",
+        "hdfs://1.2.3.4/dir",
+        LocationUtil.posixNormalize("hdfs://1.2.3.4/dir/"));
+    Assert.assertEquals(
+        "Must change // to / to enforce posix path",
+        "hdfs://1.2.3.4/foo/bar",
+        LocationUtil.posixNormalize("hdfs://1.2.3.4/foo//bar"));
+    Assert.assertEquals(
+        "Must resolve .. to previous directory to enforce posix path",
+        "hdfs://1.2.3.4/bar",
+        LocationUtil.posixNormalize("hdfs://1.2.3.4/foo/../bar"));
+    Assert.assertEquals(
+        "Must resolve . to current directory to enforce posix path",
+        "hdfs://1.2.3.4/foo/bar",
+        LocationUtil.posixNormalize("hdfs://1.2.3.4/foo/.//bar"));
+  }
+
+  @Test
+  public void testPosixNormalizePathsWithSchemeAndWithoutAuthority() {
+    Assert.assertEquals(
+        "Must work with the root directory representation",
+        "file:///",
```

Comment:
This is a place where Hadoop Path differs, it resolves to `file:/`, but that seems like an invalid resolution.

Also I am treating `/` as a special case, both here and in a path without a scheme. If we don't do this, then a user defining a table root location `/` will be converted to empty path and all locations become relative.